include/grpc/support/dynamic_annotations.h

```diff
@@ -0,0 +1,67 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_SUPPORT_DYNAMIC_ANNOTATIONS_H
+#define GRPC_SUPPORT_DYNAMIC_ANNOTATIONS_H
+
+#include <grpc/support/port_platform.h>
+
+#ifdef GRPC_TSAN_ENABLED
+
+#define TSAN_ANNOTATE_HAPPENS_BEFORE(addr) \
+  AnnotateHappensBefore(__FILE__, __LINE__, (void*)(addr))
+#define TSAN_ANNOTATE_HAPPENS_AFTER(addr) \
+  AnnotateHappensAfter(__FILE__, __LINE__, (void*)(addr))
+#define TSAN_ANNOTATE_RWLOCK_CREATE(addr) \
+  AnnotateRWLockCreate(__FILE__, __LINE__, (void*)(addr))
+#define TSAN_ANNOTATE_RWLOCK_DESTROY(addr) \
+  AnnotateRWLockDestroy(__FILE__, __LINE__, (void*)(addr))
+#define TSAN_ANNOTATE_RWLOCK_ACQUIRED(addr, is_w) \
+  AnnotateRWLockAcquired(__FILE__, __LINE__, (void*)(addr), (is_w))
+#define TSAN_ANNOTATE_RWLOCK_RELEASED(addr, is_w) \
+  AnnotateRWLockReleased(__FILE__, __LINE__, (void*)(addr), (is_w))
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+void AnnotateHappensBefore(const char* file, int line, const volatile void* cv);
```

Comment:
Where are these functions defined?  Are they provided by tsan itself?  If so, is there a header that they are defined in that we can include instead of declaring them ourselves?

Can you point me at documentation for any of this?

Reply:
That's a great question. So, there is not much document on this, but I'm vaguely familiar with the history. These functions are historically from Helgrind, that would basically use the following state machine to detect races:
http://valgrind.org/docs/manual/hg-manual.html#hg-manual.data-races.algorithm

These used to be weak symbols and linked by valgrind.

tsan used the same annotations and mechanisms, basically using identical function calls, but with a nice idea. These functions are translated into tsan instructions at compiled time and removed completely by clang and gcc when compiled using tsan. In other words, they don't have to be implemented, and unfortunately they don't have any headers to include.

Some libraries provide a default empty implementation for these functions in case they wanted to change the implementation at link time. I didn't add those because I thought they are not necessary, but if you think that would help readability, I'm happy to add those. Please let me know.